### PR TITLE
Docker release

### PR DESCRIPTION
**Pull request by daveter9**
Add `docker-release` task which will build jars, create docker images and uploads them to docker hub.
Requires docker running remote api.

### CODE REVIEW

**Comment by daveter9** [comment]
please squash this while merging.

**Comment by daveter9** [comment]
I created a workflow which triggers on tag as well, I do not think it is that useful those since it requires credentials to be stored in the grade properties and a remote docker instance. Could be updated to install docker in pipeline, use version from tag and get credentials from somewhere?

**Comment by nickyu42** [comment]
> I created a workflow which triggers on tag as well, I do not think it is that useful those since it requires credentials to be stored in the grade properties and a remote docker instance. Could be updated to install docker in pipeline, use version from tag and get credentials from somewhere?

I think the best would be to use the [Publish docker](https://github.com/marketplace/actions/publish-docker) action and get the credentials from gh secrets, but that would require creating another docker hub account and more setup. Idk if it's worth the hassle at the moment.

**Comment by daveter9** [comment]
update to ci secrets
`DOCKER_HOST` env

**Comment by daveter9** [comment]
docker new syntax

**Comment by daveter9** [comment]
`docker run -it --rm -v ${PWD}/pluginmanager.yml:/root/config.yml daveter9/hyperion-pluginmanager:0.1.0`
is now correct syntaxt

File: build.gradle.kts [hunk 1]
```diff
@@ -68,6 +73,79 @@ subprojects {
...
+                              ":pipeline:plugins:loadbalancer:", ":pipeline:plugins:pathextractor:",
+                              ":pipeline:plugins:printer:", ":pipeline:plugins:rate:",
+                              ":pipeline:plugins:reader:", ":pipeline:plugins:renamer:",
+                              ":pipeline:plugins:stresser:", ":pipeline:plugins:versiontracker:")
+
+tasks.register<DockerBuildImage>("docker-image") {
+    group = "docker"
+    description = "Create docker image"
+    doFirst {
+        println("building image")
```
Comment (molenzwiebel) [inline build.gradle.kts:87]: remove

File: build.gradle.kts [hunk 2]
```diff
@@ -68,6 +73,79 @@ subprojects {
...
+    }
+}
+
+tasks.register<DockerPushImage>("docker-push") {
+    group = "docker"
+    description = "Pushed a docker image to docker hub"
+    doFirst {
+        println("pushing image")
```
Comment (molenzwiebel) [inline build.gradle.kts:95]: remofve

File: build.gradle.kts [hunk 3]
```diff
@@ -68,6 +73,79 @@ subprojects {
...
+    }
+}
+
+configure(subprojects.filter { releaseArtifacts.contains<String>(it.path + ":")}) {
```
Comment (molenzwiebel) [inline build.gradle.kts:99]: mismatched spacing

File: build.gradle.kts [hunk 4]
```diff
@@ -68,6 +73,79 @@ subprojects {
...
+    version = (if (project.properties["version"] != "unspecified") { project.properties["version"] } else { "0.1.0" })!!
```
Comment (molenzwiebel) [inline build.gradle.kts:100]: should prob not hardcode this to 0.1.0 but instead error

File: build.gradle.kts [hunk 5]
```diff
@@ -68,6 +73,79 @@ subprojects {
...
+    val artifactName = "hyperion-${artifactParts.joinToString("-", "", "", limit = -1)}:$version"
+
+    val imageName = "$dockerId/$artifactName"
+
+    tasks.register<DockerBuildImage>("docker-image") {
+        dependsOn(artifact + "shadowJar")
+
+        docker {
+            url.set(project.properties["dockerHost"].toString())
+            //dockerFile.set(dockerfileArtifactPath)
```
Comment (molenzwiebel) [inline build.gradle.kts:114]: still needed?

File: build.gradle.kts [hunk 6]
```diff
@@ -68,6 +73,79 @@ subprojects {
...
+    group = "docker"
+    description = "Pushed a docker image to docker hub"
+    doFirst {
+        println("pushing image")
+    }
+}
+
+configure(subprojects.filter { releaseArtifacts.contains<String>(it.path + ":")}) {
+    version = (if (project.properties["version"] != "unspecified") { project.properties["version"] } else { "0.1.0" })!!
+    val dockerId = "daveter9"
```
Comment (molenzwiebel) [inline build.gradle.kts:101]: should probably change this (use the env variable?)

File: build.gradle.kts [hunk 7]
```diff
@@ -68,6 +73,79 @@ subprojects {
...
+                username.set(project.properties["dockerUsername"].toString())
+                password.set(project.properties["dockerPassword"].toString())
+                email.set(project.properties["dockerEmail"].toString())
+            }
+        }
+    }
+
+    tasks.register<DockerPushImage>("docker-push") {
+        dependsOn(artifact + "docker-image")
+        //println("docker image $project")
```
Comment (molenzwiebel) [inline build.gradle.kts:128]: remove

File: aggregator/Dockerfile [hunk 8]
```diff
@@ -1,3 +1,10 @@
...
+# first build the jar
+# assumses build context is the root of this project
+FROM gradle:jdk14 as builder
+COPY . .
+RUN gradle aggregator:shadowJar
+
+# run the built jar
+FROM openjdk:14 AS runner
+COPY --from=builder /home/gradle/aggregator/build/aggregator-all.jar .
+CMD java -jar aggregator-all.jar
```
Comment (molenzwiebel) [inline aggregator/Dockerfile:10]: Does this not need a config file? Should we at least `touch` the config file so we can use the launch shorthand?

File: build.gradle.kts [hunk 9]
```diff
@@ -39,6 +42,7 @@ allprojects {
 
 subprojects {
     apply(plugin = "kotlin")
+    apply(plugin = "com.bmuschko.docker-remote-api")
```
Comment (molenzwiebel) [inline build.gradle.kts:45]: I don't think we need this any more?

File: build.gradle.kts [hunk 10]
```diff
@@ -1,4 +1,7 @@
+import com.bmuschko.gradle.docker.tasks.image.DockerBuildImage
+import com.bmuschko.gradle.docker.tasks.image.DockerPushImage
```
Comment (molenzwiebel) [inline build.gradle.kts:2]: I don't think we need this any more?

File: pipeline/plugins/adder/README.md [hunk 11]
```diff
@@ -82,7 +82,14 @@ java -jar build/adder-all.jar [path to config]
 
 If you have a built version of this plugin, you can use the accompanied Dockerfile to set up an image with all pre-requisites installed. Please note that this Dockerfile will not compile the plugin for you.
 
-Please note that the docker container for this plugin will load the configuration file from the `CONFIGFILE` environment variable, and not the command line arguments.
+Please note that the docker container for this plugin will load the configuration file from `/root/config.yml` in its container.
+
+A pre-built image is available at dockerhub under `daveter9/hyperion-pipeline-plugins-adder:0.1.0`.
```
Comment (molenzwiebel) [inline pipeline/plugins/adder/README.md:87]: Update this with the new repository path?

File: pipeline/plugins/adder/README.md [hunk 12]
```diff
@@ -82,7 +82,14 @@ java -jar build/adder-all.jar [path to config]
 
 If you have a built version of this plugin, you can use the accompanied Dockerfile to set up an image with all pre-requisites installed. Please note that this Dockerfile will not compile the plugin for you.
```
Comment (molenzwiebel) [inline pipeline/plugins/adder/README.md:83]: We should probably update this, as it now _does_ compile the plugin for you. Should probably also add a note that it assumes the build context is the root of the repository.

File: aggregator/Dockerfile (hunk 8 above)
Comment (molenzwiebel) [reply aggregator/Dockerfile:10]:
```
RUN touch /root/config.yml
ENV HYPERION_AGGREGATOR_CONFIG=/root/config.yml
```

**Pull request merged**